keyboards/molecule/adns.c

```diff
@@ -0,0 +1,266 @@
+}
+
+void pointing_device_init(void) {
+    dprint("STARTING INTI\n");
+
+    spi_init();
+    // reset serial port
+    adns_begin();
+    adns_end();
+
+    // reboot
+    adns_write(REG_Power_Up_Reset, 0x5a);
+    wait_ms(50);
+    // read registers and discard
+    adns_read(REG_Motion);
+    adns_read(REG_Delta_X_L);
+    adns_read(REG_Delta_X_H);
+    adns_read(REG_Delta_Y_L);
+    adns_read(REG_Delta_Y_H);
+
+    // upload firmware
+
+    // set the configuration_IV register in 3k firmware mode
+    // bit 1 = 1 for 3k mode, other bits are reserved
+    adns_write(REG_Configuration_IV, 0x02);
+
+    // write 0x1d in SROM_enable reg for initializing
+    adns_write(REG_SROM_Enable, 0x1d);
+
+    // wait for more than one frame period
+    // assume that the frame rate is as low as 100fps... even if it should never be that low
+    wait_ms(10);
+
+    // write 0x18 to SROM_enable to start SROM download
+    adns_write(REG_SROM_Enable, 0x18);
+
+    // write the SROM file (=firmware data)
+
+    // write burst destination adress
+    adns_begin();
+    spi_write(REG_SROM_Load_Burst | 0x80);
+    wait_us(15);
+    // send all bytes of the firmware
+    unsigned char c;
+    for(int i = 0; i < firmware_length; i++){
+        c = (unsigned char)pgm_read_byte(firmware_data + i);
+        spi_write(c);
+        wait_us(15);
+    }
+
+    adns_end();
+    wait_ms(10);
+
+    // enable laser(bit 0 = 0b), in normal mode (bits 3,2,1 = 000b)
+    // reading the actual value of the register is important because the real
+    // default value is different from what is said in the datasheet, and if you
+    // change the reserved bytes (like by writing 0x00...) it would not work.
+    uint8_t laser_ctrl0 = adns_read(REG_LASER_CTRL0);
+    adns_write(REG_LASER_CTRL0, laser_ctrl0 & 0xf0);
+
+    wait_ms(1);
+
+    // set the configuration_I register to set the CPI
+    // 0x01 = 50, minimum
+    // 0x44 = 3400, default
+    // 0x8e = 7100
+    // 0xA4 = 8200, maximum
+    adns_write(REG_Configuration_I, 0x10);
+
+    wait_ms(100);
+    dprint("INIT ENDED\n");
+}
+
+int16_t convertDeltaToInt(uint8_t high, uint8_t low){
+
+    // join bytes into twos compliment
+    //int16_t twos_comp = (high << 8) | low;
+    //return twos_comp;
+    return (high << 8) | low;
+}
+
+void readSensor(void) {
+
+    adns_begin();
+
+    // read from Motion_Burst to enable burt mode
+    spi_write(REG_Motion_Burst & 0x7f);
+
+    // Wait one frame per docs, thanks u/kbjunky
+    wait_us(100);
+    uint8_t burst_data[pixel_sum];
+
+    for (int i = 0; i < pixel_sum; ++i) {
+        burst_data[i] = spi_read();
+    }
+
+    delta_x += convertDeltaToInt(burst_data[delta_x_h], burst_data[delta_x_l]);
+    delta_y += convertDeltaToInt(burst_data[delta_y_h], burst_data[delta_y_l]);
+    // Only consider the MSB for motion as this byte has other status bits
+    motion_ind = burst_data[motion] & 0b10000000;
+    adns_end();
+
+}
+void pointing_device_task(void) {
+    readSensor();
```

Comment:
If you have this return an x and y coordinate, you wouldn't need the global delta_x and delta_y report.   And the "send" function will clear the report, after it's sent.

Reply:
Ok. I modify the function to return delta reports.